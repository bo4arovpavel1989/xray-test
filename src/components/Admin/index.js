import React from 'react'
import { Link, withRouter } from 'react-router-dom'
import { getData, deleteData, postData, postFile, downloadFile } from './../../actions'

class Admin extends React.Component {
  constructor (props) {
    super(props)

    this.state = {
      questions: [],
      settings: [],
      tests: [],
      submitting: false,
      err: false
    }

    this.deleteObj = this.deleteObj.bind(this);
    this.refreshTest = this.refreshTest.bind(this);
    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
    this.saveDb = this.saveDb.bind(this);
    this.handleDbRecovery = this.handleDbRecovery.bind(this);

    this.getData = this.props.getData || getData;
    this.postData = this.props.postData || postData;
    this.postFile = this.props.postFile || postFile;
    this.deleteData = this.props.deleteData || deleteData;
    this.downloadFile = this.props.downloadFile || downloadFile;
  }

  componentDidMount () {
    this.getSettings();
    this.getTests();
    this.getQuestions();
  }

  getSettings () {
    return this.getData('settings')
               .then(settings => this.setState({ settings }))
               .catch(err => this.setState({ err: true }))
  }

  getTests () {
    return this.getData('tests')
               .then(tests => this.setState({ tests }))
               .catch(err => this.setState({ err: true }))
  }

  getQuestions () {
    return this.getData('questions')
               .then(questions => this.setState({ questions }))
               .catch(err => this.setState({ err: true }))
  }

  getTestData () {
    this.getTests();
    this.getQuestions();
  }

  deleteObj (obj, name) {
    let conf = window.confirm('Вы уверены?');

    if (conf) {
      return this.deleteData(`deleteobj/${obj}/${name}`)
                 .then(rep => this.getTestData())
                 .catch(err => this.setState({ err: true }))
    }
  }

  refreshTest (name) {
    return this.postData('test', { name })
               .then(rep => window.alert('Успешно обновлено!'))
               .catch(window.alert)
  }

  handleChange (e) {
    let { value, type } = e.target;
    const { description } = e.target.dataset;
    const name = e.target.id;
    const { settings } = this.state;

    value = this.handleTypes(value, type);

    settings.forEach((tune, i) => {
      if (name === tune.name) settings[i] = { name, value, type, description }
    })

    this.setState({ settings })
  }

  handleTypes (val, type) {
    const typesMap = {
      number: () => Number(val)
    }

    if (typesMap[type]) return typesMap[type](val)
    else return val
  }

  saveDb () {
    this.setState({ submitting: true })

    return this.getData('savedb')
               .then(() => {
                  // Added timeout, otherwise firefox downloads only 1 file
                  this.downloadFile('download_dump/tests.json', 'tests.json');
                  setTimeout(() =>
                    this.downloadFile('download_dump/questions.json', 'questions.json'),
                  100);
               })
               .catch(window.alert)
               .finally(() => this.setState({ submitting: false }))
  }

  handleDbRecovery (e) {
    e.preventDefault();

    let sure = window.confirm('Вы уверены? Текущая база данных будет перезаписана!')

    if (!sure) return;

    this.setState({ submitting: true })

    const data = this.props.handleFormData || this.handleFormData()

    return this.postFile('loaddb', data)
               .then(() => {
                  window.alert('Успешно восстановлено!');
                  this.getTestData();
               })
               .catch(window.alert)
               .finally(() => this.setState({ submitting: false }))
  }

  handleFormData () {
    const questions = document.querySelector('input[type="file"][name="questions"]');
    const tests = document.querySelector('input[type="file"][name="tests"]');
    const data = new FormData();

    data.append('questions', questions.files[0]);
    data.append('tests', tests.files[0]);

    return data;
  }

  handleSubmit (e) {
    e.preventDefault();
    this.setState({ submitting: true })

    const { settings } = this.state;

    return postData('settings', { settings })
            .then(rep => window.alert('Успешно сохранено!'))
            .catch(window.alert)
            .finally(() => this.setState({ submitting: false }))
  }

  render () {
    const { settings, tests, questions, submitting } = this.state;

    return (
      <div className='container'>
        <div className='newTest'>
          <div className='menuItem'>
            <Link className='menuButton' to='/create/test'>Создать новый тест</Link>
            <Link className='menuButton' to='/create/question'>Создать новый вопрос</Link>
          </div>
          <div className='dbControls'>
            <button disabled={submitting} className='menuButton small saveDb' onClick={this.saveDb}>Сохранить базу</button>
            <form className='loadDb' onSubmit={this.handleDbRecovery}>
              <label> Выберите JSON файл с вопросами &emsp;
                <input type='file' name='questions' accept='.json' required/>
              </label><br/>
              <label> Выберите JSON файл с тестами &emsp;
                <input type='file' name='tests' accept='.json' required/>
              </label><br/>
              <input disabled={submitting} type='submit' value='Восстановить базу'/>
            </form>
          </div>
        </div>
          <h2>Настройки</h2>
          <div className='formArea'>
            <form onSubmit={this.handleSubmit}>
              {
                settings.map(tune => {
                  return <div key={tune._id}>
                    <label>
                      {tune.description}: &emsp;
                      <input
                        id={tune.name}
                        data-description={tune.description}
                        type={tune.type}
                        value={tune.value}
                        onChange={this.handleChange}
                      />
                    </label>
                  </div>
                })
              }
              <div>
                <input disabled={submitting} type='submit' value='Сохранить'/>
              </div>
            </form>
          </div>
        <div>
        </div>
        <div className='oldTest'>
          <h2>Тесты</h2>
          <ul>
            {
              tests.map(test => {
                return <li key={test._id}>{test.name} &emsp;
                <a onClick={() => this.refreshTest(test.name)}>Обновить</a> &emsp;
                <a className='danger' onClick={() => this.deleteObj('Test', test.name)}>Удалить</a>
                </li>
              })
            }
          </ul>
        </div>
        <div className='oldQuestions'>
          <h2>Вопросы</h2>
          <ul>
            {
              questions.map(q => {
                return <li key={q._id}>{q.name} &emsp;
                  <Link to={`/create/question?question=${q.name}`}>Изменить</Link>&emsp;
                  <a className='danger' onClick={() => this.deleteObj('Question', q.name)}>Удалить</a>
                </li>
              })
            }
          </ul>
        </div>
      </div>
    )
  }
}

export default withRouter(Admin)
export { Admin }
