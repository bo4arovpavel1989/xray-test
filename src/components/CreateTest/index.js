import React from 'react'
import { withRouter } from 'react-router-dom'
import { postData } from './../../actions'

class CreateTest extends React.Component {
  constructor () {
    super()

    this.state = {
      err: false,
      loading: false,
      name: ''
    }

    this.handleChange = this.handleChange.bind(this);
    this.handleSubmit = this.handleSubmit.bind(this);
  }

  handleSubmit (e) {
    e.preventDefault();

    const { name } = this.state;

    this.setState({ loading: true });

    return postData('test', { name })
            .then(rep => window.alert('Успешно сохранено!'))
            .catch(err => window.alert(err))
            .finally(() => this.setState({ loading: false }))
  }

  handleChange (e) {
    const data = {};

    data[e.target.id] = e.target.value;

    this.setState(data);
  }

  render () {
    const { loading } = this.state;

    return (
      <div className='container'>
        <div className='formArea'>
          <form onSubmit={this.handleSubmit}>
            <div>
              <span>
                <label>
                  Номер теста: &emsp;<input id='name' onChange={this.handleChange} type='number' min='1'/>
                </label>
              </span>
              <span>
                <input disabled= { loading } type='submit' value='Создать'/>
              </span>
            </div>
          </form>
        </div>
      </div>
    )
  }
}

export default withRouter(CreateTest)
