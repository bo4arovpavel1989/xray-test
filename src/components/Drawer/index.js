class Drawer {
  constructor (canvas) {
    this.canvas = canvas;
    this.ctx = this.canvas.getContext('2d');

    this.zones = [];

    this.x1 = 0;
    this.x2 = 0;
    this.y1 = 0;
    this.y2 = 0;

    this.prevProperties = [0, 0, 0, 0];

    this.clearListeners();
    this.isDrawing = false;

    this.handelMouseUp = this.handelMouseUp.bind(this);
    this.handelMouseMove = this.handelMouseMove.bind(this);
    this.handelMouseDown = this.handelMouseDown.bind(this);
    this.clearZones = this.clearZones.bind(this);
  }

  handelMouseDown (e) {
    this.isDrawing = true;

    this.x1 = e.offsetX;
    this.y1 = e.offsetY;
  }

  handelMouseMove (e) {
    if (this.isDrawing) {
      this.clearPreviousRect();

      this.x2 = e.offsetX;
      this.y2 = e.offsetY;

      const rectProperties = this.getRectProperties();

      this.drawRect(rectProperties);
      this.prevProperties = rectProperties;
    }
  }

  handelMouseUp (e) {
    this.isDrawing = false;

    this.clearPreviousRect();

    this.x2 = e.offsetX;
    this.y2 = e.offsetY;

    const rectProperties = this.getRectProperties();

    this.drawRect(rectProperties);
    this.prevProperties = [0, 0, 0, 0];
    this.saveZones();
  }

  drawRect (rectProperties, color = 'red') {
    this.ctx.strokeStyle = color;
    this.ctx.rect(...rectProperties);
    this.ctx.stroke();
  }

  clearPreviousRect () {
    const [x, y, width, heigth] = this.prevProperties;

    this.ctx.clearRect(x - 1, y - 1, width + 2, heigth + 2);
    this.ctx.beginPath();
  }

  saveZones () {
    const leftUpCorner = this.getLeftUpCorner();
    const rightDownCorner = this.getRightDownCorner();

    this.zones.push([...leftUpCorner, ...rightDownCorner]);
  }

  /**
   * Method draws rectangles got from API
   */
  drawOldZones (rectArray) {
    this.zones = rectArray;

    rectArray.forEach(dimensions => {
      const rectProperties = [dimensions[0], dimensions[1]];
      const width = dimensions[2] - dimensions[0];
      const height = dimensions[3] - dimensions[1];

      rectProperties.push(width);
      rectProperties.push(height);

      this.drawRect(rectProperties);
    })
  }

  getLeftUpCorner () {
    const { x1, x2, y1, y2 } = this;

    return [Math.min(x1, x2), Math.min(y1, y2)]
  }

  getRightDownCorner () {
    const { x1, x2, y1, y2 } = this;

    return [Math.max(x1, x2), Math.max(y1, y2)]
  }

  getRectProperties () {
    const leftUpCorner = this.getLeftUpCorner();
    const rightDownCorner = this.getRightDownCorner();

    let width = rightDownCorner[0] - leftUpCorner[0];
    let height = rightDownCorner[1] - leftUpCorner[1];

    return [...leftUpCorner, width, height];
  }

  start () {
    this.canvas.addEventListener('mousedown', this.handelMouseDown)
    this.canvas.addEventListener('mouseup', this.handelMouseUp)
    this.canvas.addEventListener('mousemove', this.handelMouseMove)
  }

  clearListeners () {
    this.canvas.removeEventListener('mousedown', this.handelMouseDown)
    this.canvas.removeEventListener('mouseup', this.handelMouseUp)
    this.canvas.removeEventListener('mousemove', this.handelMouseMove)
  }

  getZones () {
    return this.zones;
  }

  clearZones () {
    this.zones = [];
    this.ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
    this.ctx.beginPath();
  }

  reset () {
    this.clearListeners();
    this.clearZones();
  }
}

export default Drawer;
