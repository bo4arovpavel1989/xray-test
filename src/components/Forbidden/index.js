import React from 'react'
import { withRouter } from 'react-router-dom'

class Forbidden extends React.Component {
  render () {
    return (
      <div>
        <h1>403 Нет доступа</h1>
        <h3>Залогиньтесь для продолжения</h3>
      </div>
    )
  }
}

export default withRouter(Forbidden)
