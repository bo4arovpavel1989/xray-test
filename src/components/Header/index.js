import React from 'react';
import {Link, withRouter} from 'react-router-dom';
import PropTypes from 'prop-types';
import Login from './Login';
import './Header.sass';

const Header = props=>{
  const {isAdmin, logoff, login} = props;

  return (
    <div className='headerArea'>
      <Link className='header' to='/'>
        <h1>XRay-test</h1>
      </Link>
      {
        isAdmin ?
          <div className='loginArea'>
            <a className='logoff' onClick={logoff}>
              Выйти
            </a>
          </div>
          :
          <Login login={login}/>
        }
    </div>
  )
};

Header.propTypes = {
  isAdmin: PropTypes.bool.isRequired,
  logoff: PropTypes.func.isRequired,
  login: PropTypes.func.isRequired
};

export default withRouter(Header);
