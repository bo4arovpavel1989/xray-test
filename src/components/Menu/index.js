import React from 'react'
import { withRouter } from 'react-router-dom'

class Menu extends React.Component {
  constructor () {
    super()
  }

  render () {
    return (
      <div>
      </div>
    )
  }
}

export default withRouter(Menu)
