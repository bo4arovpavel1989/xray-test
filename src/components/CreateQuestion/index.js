import React from 'react'
import { withRouter } from 'react-router-dom'
import { postFile, postData, getData } from './../../actions'
import Drawer from '../Drawer'
import './CreateQuestion.sass'

class CreateQuestion extends React.Component {
  constructor () {
    super()

    this.state = {
      loading: false,
      err: false,
      imgPath: '',
      dimensions: { width: 0, height: 0 },
      dangerZones: [],
      name: '',
      isDanger: '1'
    }

    this.handleChange = this.handleChange.bind(this);
    this.uploadPhotos = this.uploadPhotos.bind(this);
    this.clearCanvas = this.clearCanvas.bind(this);
    this.saveQuestion = this.saveQuestion.bind(this);
  }

  componentDidMount () {
    const canvasDraw = '#canvasDrawArea';
    const canvasBackground = '#canvasBackground';
    this.setState({ canvasBackground, canvasDraw })

    this.drawer = new Drawer(document.querySelector(canvasDraw));
    this.checkIfQuestionCreated()
  }

  componentWillUnmount () {
    this.drawer.reset();
  }

  /**
   * Get selector for canvas draw layer
   * @returns {Object} - DOM selector
   */
  getCanvasDraw () {
    return document.querySelector(this.state.canvasDraw)
  }

  /**
   * Get selector for canvas background layer
   * @returns {Object} - DOM selector
   */
  getCanvasBackground () {
    return document.querySelector(this.state.canvasBackground)
  }

  /**
   * Method checks location query
   * if query is defined ( /create/question?question=name )
   * it means user want to edit old question
   * so we fetch question to API
   * @return {Function} - fetch to API for question data
   */
  checkIfQuestionCreated () {
    if (this.props.location.search.length > 0) {
      const query = new URLSearchParams(this.props.location.search);
      const question = query.get('question');

      return this.getQuestion(question)
    }
  }

  /**
   * Method fetches to API for question data
   * if any - renders it
   * @returns {Function} - fetch to API
   */
  getQuestion (question) {
    return getData(`question/${question}`)
            .then(rep => {
              console.log(rep)
              if (rep) {
                this.setState(rep, this.prepareCanvas)
              }
            })
            .catch(err => window.alert(err))
  }

  /**
   * Method handles upload of slide and photo
   * then it gets dimensions of slide and prepares canvas for it
   * @return {void}
   */
  uploadPhotos (e) {
    e.preventDefault();

    const data = this.handleFormData();

    this.setState({ loading: true });
    postFile('preupload', data)
      .then(rep => {
        const { name } = this.state;
        const imgPath = `/images/${name}_slide.${rep.type}`;
        this.drawer.reset();

        this.setState({
          loading: false,
          imgPath: imgPath,
          dimensions: rep
        }, this.prepareCanvas)
      })
      .catch(err => {
        this.setState({ loading: false })
        window.alert(err)
      });
  }

  /**
   * Method gatheres all formData to single object for upload
   * @return {Object} - object to upload
   */
  handleFormData () {
    const slide = document.querySelector('input[name="slide"]');
    const photo = document.querySelector('input[name="photo"]');
    const { name, isDanger } = this.state;
    const data = new FormData();

    data.append('question', name);
    data.append('slide', slide.files[0]);
    if (isDanger === '1') data.append('photo', photo.files[0]);

    return data;
  }

  /**
   * Method handles changes in input and pass it to state
   * @returns {void}
   */
  handleChange (e) {
    const data = {};

    data[e.target.id] = e.target.value;

    this.setState(data);
  }

  /**
   * Method sets canvas width and heigth, add background image
   * and inits drawing dangerzone rectangels
   * @returns {void}
   */
  prepareCanvas () {
    this.setCanvasDimensions();

    const background = new Image();

    background.src = this.state.imgPath;
    background.onload = () => {
      this.setCanvasBackground(background)
      this.drawDangerZone();
    }
  }

  /**
   * Method set canvas width and heigth - both for background and draw layers
   * @returns {void}
   */
  setCanvasDimensions () {
    const { width, height } = this.state.dimensions;

    this.getCanvasDraw().width = width;
    this.getCanvasDraw().height = height;
    this.getCanvasBackground().width = width;
    this.getCanvasBackground().height = height;
  }

  /**
   * Method calls background layer canvas to draw background
   * @returns {Function} - drawImage method of canvas
   */
  setCanvasBackground (bg) {
    return this.getCanvasBackground().getContext('2d').drawImage(bg, 0, 0)
  }

  /**
   * Method draws old rectangles got from API if any
   * and starts to listen to new user draw
   * @returns {Function} - start method of this.draw
   */
  drawDangerZone (c) {
    const { dangerZones } = this.state;

    if (dangerZones.length > 0) this.drawer.drawOldZones(dangerZones);

    return this.drawer.start()
  }

  /**
   * Method calls clearZones mthod of this.drawer
   * to clear canvas draw layer
   */
  clearCanvas () {
    return this.drawer.clearZones()
  }

  /**
   * Method gatheres all question data to single object and post to API
   * @returns {Function} - fetch post API
   */
  saveQuestion () {
    const { imgPath, name, isDanger, dimensions } = this.state;

    const question = {
      name,
      isDanger,
      dangerZones: this.drawer.getZones(),
      imgPath,
      dimensions
     };

     this.setState({ loading: true });

     return postData('savequestion', question)
              .then(rep => {
                window.alert('Успешно сохранено!');
                this.setState({ loading: false })
              })
              .catch(err => {
                window.alert(err);
                this.setState({ loading: false })
              })
  }

  render () {
    const { loading, isDanger, name } = this.state;

    return (
      <div className='container'>
        <div className='formArea'>
          <form onSubmit={this.uploadPhotos}>
            <div>
              <label>
                Вопрос: &emsp;
                <input
                  id='name'
                  onChange={this.handleChange}
                  pattern='[0-9]+_[0-9]+'
                  name='name'
                  type='text'
                  value = {name.length > 0 ? name : ''}
                  placeholder='№ Теста_№ вопроса'
                  required
                />
              </label>
            </div>
            <div>
              <label>
                Загрузите слайд: &emsp;
                <input
                  name='slide'
                  type='file'
                  required
                />
              </label>
            </div>
            <div>
              <label>
                Багаж опасен? &emsp;
                <select
                  id='isDanger'
                  onChange={this.handleChange}
                  value={isDanger === '1' ? '1' : '0'}
                >
                  <option value='1'>Да</option>
                  <option value='0'>нет</option>
                </select>
              </label>
            </div>
            {
              isDanger === '1' ?
              <div>
                <label>
                  Загрузите фото: &emsp;
                  <input
                    name='photo'
                    type='file'
                    required
                  />
                </label>
              </div> :
              <span></span>
            }
            <div>
              <input disabled={loading} type='submit' value='Продолжить'/>
            </div>
          </form>
        </div>
        <div>
          <button onClick={this.clearCanvas}>Очистить</button>
          <button disabled={loading} onClick={this.saveQuestion}>Сохранить</button>
          <div className='canvasArea'>
            <canvas id="canvasBackground"></canvas>
            <canvas id="canvasDrawArea"></canvas>
          </div>
        </div>
      </div>
    )
  }
}

export default withRouter(CreateQuestion)
