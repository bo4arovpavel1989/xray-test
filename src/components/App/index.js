import React from 'react'
import Header from '../Header'
import Menu from '../Menu'
import Test from '../Test'
import Admin from '../Admin'
import Forbidden from '../Forbidden'
import CreateTest from '../CreateTest'
import CreateQuestion from '../CreateQuestion'
import { BrowserRouter, Route, Switch } from 'react-router-dom'
import '../../main.sass'
import { postData } from '../../actions.js'
import { setToken } from '../../helpers'

class App extends React.Component {
  constructor () {
    super()

    this.state = {
      isAdmin: false,
      authFail: false,
      isLogging: false,
      token: undefined
    }

    this.login = this.login.bind(this)
    this.logoff = this.logoff.bind(this)
    this.handleAuth = this.handleAuth.bind(this)
  }

  login (e) {
    e.preventDefault()

    const data = {
      login: e.target.login.value,
      password: e.target.password.value
    }

    this.setState({ isLogging: true })

    postData('login', data)
      .then(this.handleAuth)
      .catch(console.log)
      .finally(() => this.setState({ isLogging: false }))
  }

  handleAuth (rep) {
    if (rep.auth) setToken(rep.token)
    this.setState({ isAdmin: rep.auth, authFail: !rep.auth, token: rep.token })
  }

  logoff () {
    this.setState({
      isAdmin: false,
      token: undefined
     })

     setToken(null)
  }

  render () {
    const { isAdmin, authFail, isLogging } = this.state

    return (
      <BrowserRouter>
        <div>
          <Header
           isAdmin={isAdmin}
           authFail={authFail}
           isLogging = {isLogging}
           login={this.login}
           logoff={this.logoff}
           />
          <Switch>
            <Route path='/' render={Menu} exact/>
            <Route path='/test' render={Test}/>
            <Route path='/admin' render={isAdmin ? Admin : Forbidden}/>
            <Route path='/create/test' render={isAdmin ? CreateTest : Forbidden}/>
            <Route path='/create/question' render={isAdmin ? CreateQuestion : Forbidden}/>
          </Switch>
        </div>
      </BrowserRouter>
    )
  }
}

export default App;
