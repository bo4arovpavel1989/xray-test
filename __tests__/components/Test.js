import React from 'react';
import { MemoryRouter } from 'react-router-dom'
import { Test } from '../../src/components/Test';
import Enzyme, { mount } from 'enzyme';
import { shallowToJson } from 'enzyme-to-json';
import Adapter from 'enzyme-adapter-react-16';

Enzyme.configure({ adapter: new Adapter() });

describe('Test component', () => {
  const props = {
    getData: jest.fn((arg) => new Promise((res, rej) => {
        if (arg === 'settings') {
          res([{ name: 'test', _id: 'testID', value: 'testValue', description: 'test' }])
        }

        res([{ name: 'test', _id: 'testID' }])
     })),
     drawer: {
       reset: jest.fn(),
       start: jest.fn(),
       clearZones: jest.fn(),
       getZones: jest.fn(() => [[0, 0, 100, 100]])
     },
     prepareCanvas:  jest.fn()
  };
  const initialState = {
    loading: false,
    testStarted: false,
    testFinished: false,
    historyMode: false,
    testHistory: [],
    user: '',
    settings: { test: 'testValue' },
    tests: [{ _id: 'testID', name: 'test' }],
    currentQuestion: -1,
    questions: [],
    total: 100
  };

  const component = mount(
    <MemoryRouter initialEntries={[ { key: 'testKey' } ]}>
      <Test { ...props }/>
    </MemoryRouter>
  );


  describe('component mounted', () => {
    afterAll(() => {
      props.getData.mockClear();
    });

    it('should render correct', () => {
      expect(shallowToJson(component)).toMatchSnapshot()
    })

    it('should call 2 initial getData (tests, settings) functions', () => {
      expect(props.getData).toHaveBeenCalledTimes(2)
    })

    it('should have initial state', () => {
      expect(component.find(Test).instance().state).toEqual(initialState)
    })
  })

  describe('handle userArea submit', () => {
    beforeAll(() => {
      component.find('.userLogin #user').simulate('change', {
        target: {
          value: 'testUser',
          id: 'user' }
        });

      component.find('.userLogin').simulate('submit')
    });

    afterAll(() => {
      props.getData.mockClear();
      component.find(Test).instance().setState(initialState);
    });

    it('should call 2 initial getData (tests, settings) functions', () => {
      expect(props.getData).toHaveBeenCalledWith('questions/test');
    })

    it('should have questions in state', () => {
      expect(component.find(Test).instance().state.questions).toEqual([{ _id: 'testID', name: 'test' }])
    })

    it('should have test-started state', () => {
      expect(component.find(Test).instance().state.testStarted).toEqual(true)
    })

    it('should have currentQuestion state = 0', () => {
      expect(component.find(Test).instance().state.currentQuestion).toEqual(0)
    })

    it('should have user state = testUser', () => {
      expect(component.find(Test).instance().state.user).toEqual('testUser')
    })
  })

  describe('test started', () => {
    beforeAll(() => {
      component.find(Test).instance().setState({
        testStarted: true,
        questions: [{ name: 'name', dimensions: { width: 100, height: 100 } }],
        currentQuestion: 0
       })
    })
    afterAll(() => {
       component.find(Test).instance().setState(initialState);
    });

    it('should render correct', () => {
      expect(shallowToJson(component)).toMatchSnapshot()
    })
  })

  describe('test finished', () => {
    beforeAll(() => {
      component.find(Test).instance().setState({
        testStarted: false,
        testFinished: true,
        settings: { errorThreshold: 80 }
      });
      component.update();
    })

    afterAll(() => {
       component.find(Test).instance().setState(initialState);
    });

    it('should render correct', () => {
      expect(shallowToJson(component)).toMatchSnapshot()
    })

    it('should havel class perfect for 100% result', () => {
        expect(component.find('.resultArea h2 span').hasClass('perfect')).toEqual(true);
    })
  })
})
