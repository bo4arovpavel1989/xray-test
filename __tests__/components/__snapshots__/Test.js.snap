// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test component component mounted should render correct 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Test
      drawer={
        Object {
          "clearZones": [MockFunction],
          "getZones": [MockFunction],
          "reset": [MockFunction],
          "start": [MockFunction],
        }
      }
      getData={
        [MockFunction] {
          "calls": Array [
            Array [
              "settings",
            ],
            Array [
              "tests",
            ],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": Promise {},
            },
            Object {
              "isThrow": false,
              "value": Promise {},
            },
          ],
        }
      }
      prepareCanvas={[MockFunction]}
    >
      <div
        className="container"
      >
        <div
          className="userArea"
        >
          <withRouter(User)
            handleChange={[Function]}
            handleSubmit={[Function]}
          >
            <Route>
              <User
                handleChange={[Function]}
                handleSubmit={[Function]}
                history={
                  Object {
                    "action": "POP",
                    "block": [Function],
                    "canGo": [Function],
                    "createHref": [Function],
                    "entries": Array [
                      Object {
                        "hash": "",
                        "key": "testKey",
                        "pathname": "/",
                        "search": "",
                      },
                    ],
                    "go": [Function],
                    "goBack": [Function],
                    "goForward": [Function],
                    "index": 0,
                    "length": 1,
                    "listen": [Function],
                    "location": Object {
                      "hash": "",
                      "key": "testKey",
                      "pathname": "/",
                      "search": "",
                    },
                    "push": [Function],
                    "replace": [Function],
                  }
                }
                location={
                  Object {
                    "hash": "",
                    "key": "testKey",
                    "pathname": "/",
                    "search": "",
                  }
                }
                match={
                  Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "/",
                    "url": "/",
                  }
                }
              >
                <div>
                  <div
                    className="paranja"
                  />
                  <div
                    className="userLoginArea"
                  >
                    <form
                      className="userLogin"
                      onSubmit={[Function]}
                    >
                      <div>
                        <span>
                          <label>
                            Ваше имя:  
                            <input
                              id="user"
                              onChange={[Function]}
                              required={true}
                              type="text"
                            />
                          </label>
                        </span>
                        <span>
                          <input
                            type="submit"
                            value="Начать тест"
                          />
                        </span>
                      </div>
                    </form>
                  </div>
                </div>
              </User>
            </Route>
          </withRouter(User)>
        </div>
      </div>
    </Test>
  </Router>
</MemoryRouter>
`;

exports[`Test component test finished should render correct 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Test
      drawer={
        Object {
          "clearZones": [MockFunction],
          "getZones": [MockFunction],
          "reset": [MockFunction],
          "start": [MockFunction],
        }
      }
      getData={[MockFunction]}
      prepareCanvas={
        [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
    >
      <div
        className="container"
      >
        <div
          className="resultArea"
        >
          <div>
            <h2>
              Ваш результат:
              <span
                className="perfect"
              >
                 
                100
                %
              </span>
            </h2>
          </div>
          <div>
            <a
              onClick={[Function]}
            >
              Показать ответы
            </a>
          </div>
        </div>
      </div>
    </Test>
  </Router>
</MemoryRouter>
`;

exports[`Test component test started should render correct 1`] = `
<MemoryRouter
  initialEntries={
    Array [
      Object {
        "key": "testKey",
      },
    ]
  }
>
  <Router
    history={
      Object {
        "action": "POP",
        "block": [Function],
        "canGo": [Function],
        "createHref": [Function],
        "entries": Array [
          Object {
            "hash": "",
            "key": "testKey",
            "pathname": "/",
            "search": "",
          },
        ],
        "go": [Function],
        "goBack": [Function],
        "goForward": [Function],
        "index": 0,
        "length": 1,
        "listen": [Function],
        "location": Object {
          "hash": "",
          "key": "testKey",
          "pathname": "/",
          "search": "",
        },
        "push": [Function],
        "replace": [Function],
      }
    }
  >
    <Test
      drawer={
        Object {
          "clearZones": [MockFunction],
          "getZones": [MockFunction],
          "reset": [MockFunction],
          "start": [MockFunction],
        }
      }
      getData={[MockFunction]}
      prepareCanvas={
        [MockFunction] {
          "calls": Array [
            Array [],
            Array [],
          ],
          "results": Array [
            Object {
              "isThrow": false,
              "value": undefined,
            },
            Object {
              "isThrow": false,
              "value": undefined,
            },
          ],
        }
      }
    >
      <div
        className="container"
      >
        <div
          className="userArea"
        >
          <withRouter(User)
            handleChange={[Function]}
            handleSubmit={[Function]}
          >
            <Route>
              <User
                handleChange={[Function]}
                handleSubmit={[Function]}
                history={
                  Object {
                    "action": "POP",
                    "block": [Function],
                    "canGo": [Function],
                    "createHref": [Function],
                    "entries": Array [
                      Object {
                        "hash": "",
                        "key": "testKey",
                        "pathname": "/",
                        "search": "",
                      },
                    ],
                    "go": [Function],
                    "goBack": [Function],
                    "goForward": [Function],
                    "index": 0,
                    "length": 1,
                    "listen": [Function],
                    "location": Object {
                      "hash": "",
                      "key": "testKey",
                      "pathname": "/",
                      "search": "",
                    },
                    "push": [Function],
                    "replace": [Function],
                  }
                }
                location={
                  Object {
                    "hash": "",
                    "key": "testKey",
                    "pathname": "/",
                    "search": "",
                  }
                }
                match={
                  Object {
                    "isExact": true,
                    "params": Object {},
                    "path": "/",
                    "url": "/",
                  }
                }
              >
                <div>
                  <div
                    className="paranja"
                  />
                  <div
                    className="userLoginArea"
                  >
                    <form
                      className="userLogin"
                      onSubmit={[Function]}
                    >
                      <div>
                        <span>
                          <label>
                            Ваше имя:  
                            <input
                              id="user"
                              onChange={[Function]}
                              required={true}
                              type="text"
                            />
                          </label>
                        </span>
                        <span>
                          <input
                            type="submit"
                            value="Начать тест"
                          />
                        </span>
                      </div>
                    </form>
                  </div>
                </div>
              </User>
            </Route>
          </withRouter(User)>
        </div>
      </div>
    </Test>
  </Router>
</MemoryRouter>
`;
